bookkeeping/bookkeeper.go

```diff
@@ -0,0 +1,123 @@
+package bookkeeping
+
+import (
+	"bytes"
+	"github.com/Comcast/webpa-common/logging"
+	"github.com/go-kit/kit/log"
+	"net/http"
+	"time"
+)
+
+type CapturedResponse struct {
+	Code    int
+	Payload []byte
+	Header  http.Header
+}
+
+// RequestFunc takes the request and returns key value pairs from the request
+type RequestFunc func(request *http.Request) []interface{}
+
+// ResponseFunc takes the ResponseWriter and returns key value pairs from the request
+type ResponseFunc func(response CapturedResponse) []interface{}
+
+// Handler is for logging of the request and response
+type Handler struct {
+	next   http.Handler
+	before []RequestFunc
+	after  []ResponseFunc
+
+	logger log.Logger
+}
+
+// Option provides a single configuration option for a bookkeeping Handler
+type Option func(h *Handler)
+
+func New(options ...Option) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		if next == nil {
+			panic("next can't be nil")
+		}
+		h := &Handler{
+			next: next,
+		}
+		for _, o := range options {
+			o(h)
+		}
+		return h
+	}
+}
+
+// WithRequests take a one or many RequestFuncs to build an Option for logging key value pairs from the
+// RequestFun
+func WithRequests(requestFuncs ...RequestFunc) Option {
+	return func(h *Handler) {
+		h.before = append(h.before, requestFuncs...)
+	}
+}
+
+// WithResponses take a one or many ResponseFunc to build an Option for logging key value pairs from the
+// ResponseFunc
+func WithResponses(responseFuncs ...ResponseFunc) Option {
+	return func(h *Handler) {
+		h.after = append(h.after, responseFuncs...)
+	}
+}
+
+// WithLogger will set a custom logger for the Bookkeeping Handler
+func WithLogger(logger log.Logger) Option {
```

Comment:
Don't supply a logger via configuration.  All HTTP code should obtain the logger from the request's context, like so:

```
logging.GetLogger(request.Context()).Log(...)
```

There is an implicit default supplied by our `logging` package, so `GetLogger` never returns `nil`.